httomo/wrappers_class.py

```diff
@@ -0,0 +1,308 @@
+            xp.cuda.Device(self.gpu_id).use()  # use a particular GPU by its id
+
+    def _execute_generic(
+        self, method_name: str, params: Dict, data: xp.ndarray, reslice_ahead: bool
+    ) -> xp.ndarray:
+        """The generic wrapper to execute functions for external packages.
+
+        Args:
+            method_name (str): The name of the method to use.
+            params (Dict): A dict containing independent of httomo params.
+            data (xp.ndarray): a numpy or cupy data array.
+            reslice_ahead (bool): a bool to inform the wrapper if the reslice ahead and the conversion to numpy required.
+
+        Returns:
+            xp.ndarray: A numpy or cupy array containing processed data.
+        """
+        # set the correct GPU ID if it is required
+        if "gpu_id" in params:
+            params["gpu_id"] = self.gpu_id
+
+        if gpu_enabled:
+            if self.cupyrun:
+                # the method accepts CuPy arrays for the GPU processing
+                with xp.cuda.Device(self.gpu_id):
+                    # move the data to a particular device
+                    data = xp.asarray(data)
+            else:
+                # the method doesn't accept CuPy arrays
+                data = xp.asnumpy(data)
+
+        data = getattr(self.module, method_name)(data, **params)
+        if reslice_ahead and gpu_enabled:
+            # reslice ahead, bring data back to numpy array
+            return xp.asnumpy(data)
+        else:
+            return data
+
+    def _execute_normalize(
+        self,
+        method_name: str,
+        params: Dict,
+        data: xp.ndarray,
+        flats: xp.ndarray,
+        darks: xp.ndarray,
+        reslice_ahead: bool,
+    ) -> xp.ndarray:
+        """Normalisation-specific wrapper when flats and darks are required.
+
+        Args:
+            method_name (str): The name of the method to use.
+            params (Dict): A dict containing independent of httomo params.
+            data (xp.ndarray): a numpy or cupy data array.
+            flats (xp.ndarray): a numpy or cupy flats array.
+            darks (xp.ndarray): a numpy or darks flats array.
+            reslice_ahead (bool): a bool to inform the wrapper if the reslice ahead and the conversion to numpy required.
+
+        Returns:
+            xp.ndarray: a numpy or cupy array of the normalised data.
+        """
+        if gpu_enabled:
+            if self.cupyrun:
+                # the method accepts CuPy arrays for the GPU processing
+                with xp.cuda.Device(self.gpu_id):
+                    # move the data to a particular device
+                    data = xp.asarray(data)
+                    flats = xp.asarray(flats)
+                    darks = xp.asarray(darks)
+            else:
+                # the method doesn't accept CuPy arrays
+                data = xp.asnumpy(data)
+                flats = xp.asnumpy(flats)
+                darks = xp.asnumpy(darks)
```

Comment:
how about adding a utility method `def transfer_data(self, *args) -> tuple:` which does all the checking and conversion, to be called like:

```python
data, flats, darks = self.transfer_data(data, flats, args)
```

Reply:
yep, good idea. I can try that, thanks!